pandas/io/excel/_openpyxl.py

```diff
@@ -451,3 +462,260 @@ def write_cells(self, cells, sheet_name=None, startrow=0, startcol=0,
                             xcell = wks.cell(column=col, row=row)
                             for k, v in style_kwargs.items():
                                 setattr(xcell, k, v)
+
+
+class _OpenpyxlReader(_BaseExcelReader):
+
+    def __init__(self, filepath_or_buffer):
+        """Reader using openpyxl engine.
+
+        Parameters
+        ----------
+        filepath_or_buffer : string, path object or Workbook
+            Object to be parsed.
+        """
+        err_msg = "Install xlrd >= 1.0.0 for Excel support"
+
+        try:
+            import openpyxl
+        except ImportError:
+            raise ImportError(err_msg)
+
+        from openpyxl.cell.cell import TYPE_ERROR as CELL_TYPE_ERROR
+        self.CELL_TYPE_ERROR = CELL_TYPE_ERROR
+
+        from pandas.io.excel._base import ExcelFile
+        # If filepath_or_buffer is a url, want to keep the data as bytes so
+        # can't pass to get_filepath_or_buffer()
+        if _is_url(filepath_or_buffer):
+            filepath_or_buffer = BytesIO(_urlopen(filepath_or_buffer).read())
+        elif not isinstance(filepath_or_buffer,
+                            (ExcelFile, openpyxl.Workbook)):
+            filepath_or_buffer, _, _, _ = get_filepath_or_buffer(
+                filepath_or_buffer)
+
+        if isinstance(filepath_or_buffer, openpyxl.Workbook):
+            self.book = filepath_or_buffer
+        elif hasattr(filepath_or_buffer, "read"):
+            if hasattr(filepath_or_buffer, 'seek'):
+                filepath_or_buffer.seek(0)
+            self.book = openpyxl.load_workbook(
+                filepath_or_buffer, data_only=True)
+        elif isinstance(filepath_or_buffer, compat.string_types):
+            self.book = openpyxl.load_workbook(
+                filepath_or_buffer, data_only=True)
+        else:
+            raise ValueError('Must explicitly set engine if not passing in'
+                             ' buffer or path for io.')
+        
+
+    @property
+    def sheet_names(self):
+        return self.book.sheetnames
+
+    def get_sheet_by_name(self, name):
+        return self.book[name]
+
+    def get_sheet_by_index(self, index):
+        return self.book.worksheets[index]
+
+    def _replace_type_error_with_nan(self, rows):
+        nan = float('nan')
+        for row in rows:
+            yield [nan
+                   if cell.data_type == self.CELL_TYPE_ERROR
+                   else cell.value
+                   for cell in row]
+
+    def get_sheet_data(self, sheet, convert_float):
+        data = self._replace_type_error_with_nan(sheet.rows)
+        # TODO: support using iterator
+        # TODO: don't make strings out of data
+        return list(data)
+
+    def parse(self,
```

Comment:
So similar comment as above it would be preferable not to override this and just leave in the base class. I've noticed a vast majority of this is simply copy / paste.

Rather indifferent but if we go the route of cleanup in a follow up issue then for sure need to consolidate this as well

Reply:
The _BaseExcelReader.parse function is closely coupled to the later call of the TextParser which  makes no sense foe an openpyxl based reader as it already outputs structured, properly parsed data. Effectively the parse function for openpyxl does almost the reverse of the base parser. The base parser applies keywords to overcome limitations of xlrd and then converts the data do a dataframe. This parse functions first makes a dataframe, and then reverse applies the many keywords that the excel read supports to mimic the behaviour and pass all the tests. These fundamental differences in approach make it very difficult to keep the functions generic. The same applies to the __init__ function. This function e.g. needs the specified engine to be imported, and becomes very ugly when that is made generic, see discussion above

Reply:
So this is probably related to my earlier question about `handle_sheet_name` but would be preferable if this could be made generic in the base class. Otherwise subclasses have to do a lot more work - do you see any way to make that possible?